adr/0025-logging.md

```diff
@@ -0,0 +1,235 @@
+# 25. Logging
+
+Date: 2024-06-06
+
+## Status
+
+Proposed
+
+## Context
+
+Zarf is currently using an in-house built logging solution which in turn depends on pterm. This solution is used to output information to the end user who is using the Zarf CLI. This output is for the most part formatted with the purpose of look good for users of the CLI. Logging function calls are done both in the CLI specific code as well as in the packages. The logging is implemented in such a way that different levels exist with different impacts and log destinations. A common pattern that is used is to call `message.Fatal` whenever an error occurs. This will output the message to STDERR while writing the actual error message to a debug log and exit the program. Exiting the program in this manner makes unit testing difficult while also skipping proper context handling skipping any clean-up that was intended to run before exiting the program. Some logging components like the progress bar and spinner are accessed through a shared global state that is not thread safe. The coupling to logging becomes complicated as disabling the progress bar is a challenge while multi threading these functions which access the global state results in complicated access patterns.
+
+## Decision
+
+I am proposing to completely refactor the logging functionality to follow a more standardized format using the newish slog interfaces. On top of that we would refactor the current internationalization by converting them to standardized errors.
+
+* Replace the message package with a slog interface.
+* Replace the lang package with static errors.
+* Remove any use of `message.Fatal` and instead return errors.
+* Refactor use of `message.WarnErr` to either return error or to log the error.
+* Refactor existing functions that print formatted outputs to defer the printing to the CLI.
```

Comment:
Anything that does fancy UI things should prob happen in CMD, all lib code should use the `slog` interface

Reply:
While this is true it is not currently possible right now. The spinner is too heavily integrated into the all of the code base. The best we can do initially is to remove the need for a global spinner that has threading issues. Then in the future we can look at how we can move UI like behavior into the cmd package.

Reply:
It may not be the best pattern but an implementation to look into could be what was done with the ProgressWriter and Maru - https://github.com/defenseunicorns/maru-runner/blob/main/src/message/progress.go#L29 - that makes the spinner and progressbar a bit more generic and a bit less decoupled.

The main thing I don't love about that implementation though it that spinner's `Write()` and progressbar's `Write()` mean slightly different things, but it is probably a decent starting point for `* Define a new interface for spinner and progress bar that is passed in as a function parameter.`